diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 2a8f5d2a8cc7..f6f82f0e432b 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -7378,7 +7378,7 @@ int kvm_emulate_hypercall(struct kvm_vcpu *vcpu)
 	switch (nr) {
 		//inserted by Sangwon Song
 	case 0x283: //
-		ret = 0x0033383245504D43 // CMPE283
+		ret = 0x0033383245504D43; // CMPE283
 		break;
 	///
 	case KVM_HC_VAPIC_POLL_IRQ:
